File: app/face_detector.py
"""
Face detection module using face_recognition library (dlib-based ML).
Provides accurate face detection and landmark extraction.
"""

import face_recognition
import numpy as np
from PIL import Image
from typing import List, Tuple, Optional
import cv2


class FaceDetector:
    """ML-based face detector using dlib's HOG + CNN models."""

    def __init__(self, model: str = "hog"):
        """
        Initialize face detector.

        Args:
            model: Detection model - "hog" (faster, CPU) or "cnn" (more accurate, GPU)
        """
        self.model = model

    def detect_faces(self, image: np.ndarray) -> List[Tuple[int, int, int, int]]:
        """
        Detect all faces in an image.

        Args:
            image: RGB numpy array of the image

        Returns:
            List of face locations as (top, right, bottom, left) tuples
        """
        # face_recognition expects RGB
        if len(image.shape) == 2:
            image = cv2.cvtColor(image, cv2.COLOR_GRAY2RGB)
        elif image.shape[2] == 4:
            image = cv2.cvtColor(image, cv2.COLOR_RGBA2RGB)

        face_locations = face_recognition.face_locations(image, model=self.model)
        return face_locations

    def get_face_landmarks(self, image: np.ndarray, face_location: Tuple[int, int, int, int]) -> Optional[dict]:
        """
        Get facial landmarks for a detected face.

        Args:
            image: RGB numpy array
            face_location: (top, right, bottom, left) tuple

        Returns:
            Dictionary of facial landmarks or None
        """
        landmarks_list = face_recognition.face_landmarks(image, [face_location])
        if landmarks_list:
            return landmarks_list[0]
        return None

    def get_face_encoding(self, image: np.ndarray, face_location: Tuple[int, int, int, int]) -> Optional[np.ndarray]:
        """
        Get 128-dimensional face encoding for face matching.

        Args:
            image: RGB numpy array
            face_location: (top, right, bottom, left) tuple

        Returns:
            128-dimensional numpy array encoding
        """
        encodings = face_recognition.face_encodings(image, [face_location])
        if encodings:
            return encodings[0]
        return None

    def extract_face(self, image: np.ndarray, face_location: Tuple[int, int, int, int],
                     padding: float = 0.3) -> Tuple[np.ndarray, Tuple[int, int, int, int]]:
        """
        Extract face region from image with padding.

        Args:
            image: RGB numpy array
            face_location: (top, right, bottom, left) tuple
            padding: Padding ratio around face (0.3 = 30% extra on each side)

        Returns:
            Tuple of (face_image, padded_location)
        """
        top, right, bottom, left = face_location
        height, width = image.shape[:2]

        # Calculate padding
        face_height = bottom - top
        face_width = right - left

        pad_h = int(face_height * padding)
        pad_w = int(face_width * padding)

        # Apply padding with bounds checking
        new_top = max(0, top - pad_h)
        new_bottom = min(height, bottom + pad_h)
        new_left = max(0, left - pad_w)
        new_right = min(width, right + pad_w)

        face_image = image[new_top:new_bottom, new_left:new_right]

        return face_image, (new_top, new_right, new_bottom, new_left)


def create_face_mask(face_image: np.ndarray, landmarks: Optional[dict] = None) -> np.ndarray:
    """
    Create an elliptical mask for smooth face blending.

    Args:
        face_image: Face region as numpy array
        landmarks: Optional facial landmarks for precise masking

    Returns:
        Grayscale mask image
    """
    height, width = face_image.shape[:2]
    mask = np.zeros((height, width), dtype=np.uint8)

    # Create elliptical mask
    center = (width // 2, height // 2)
    axes = (int(width * 0.45), int(height * 0.48))

    cv2.ellipse(mask, center, axes, 0, 0, 360, 255, -1)

    # Apply Gaussian blur for smooth edges
    mask = cv2.GaussianBlur(mask, (21, 21), 11)

    return mask

File: src/faceDetector.js
/**
 * Face detection module using face-api.js (TensorFlow.js based)
 * Provides accurate face detection and landmark extraction
 */

import * as faceapi from '@vladmandic/face-api';
import * as tf from '@tensorflow/tfjs-node';
import { Canvas, Image, ImageData, createCanvas, loadImage } from 'canvas';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Patch face-api to use node-canvas
faceapi.env.monkeyPatch({ Canvas, Image, ImageData });

let modelsLoaded = false;

/**
 * Load face-api.js models
 */
export async function loadModels() {
  if (modelsLoaded) return;

  const modelsPath = path.join(__dirname, '..', 'models');

  console.log('Loading face detection models...');

  await Promise.all([
    faceapi.nets.ssdMobilenetv1.loadFromDisk(modelsPath),
    faceapi.nets.faceLandmark68Net.loadFromDisk(modelsPath),
    faceapi.nets.faceRecognitionNet.loadFromDisk(modelsPath),
  ]);

  modelsLoaded = true;
  console.log('Face detection models loaded');
}

/**
 * Detect faces in an image buffer
 * @param {Buffer} imageBuffer - Image buffer
 * @returns {Promise<Array>} Array of face detections with landmarks
 */
export async function detectFaces(imageBuffer) {
  await loadModels();

  const img = await loadImage(imageBuffer);
  const canvas = createCanvas(img.width, img.height);
  const ctx = canvas.getContext('2d');
  ctx.drawImage(img, 0, 0);

  const detections = await faceapi
    .detectAllFaces(canvas)
    .withFaceLandmarks()
    .withFaceDescriptors();

  return detections.map(d => ({
    box: {
      x: Math.round(d.detection.box.x),
      y: Math.round(d.detection.box.y),
      width: Math.round(d.detection.box.width),
      height: Math.round(d.detection.box.height),
    },
    landmarks: d.landmarks.positions.map(p => ({ x: p.x, y: p.y })),
    descriptor: Array.from(d.descriptor),
    score: d.detection.score,
  }));
}

/**
 * Detect the largest face in an image
 * @param {Buffer} imageBuffer - Image buffer
 * @returns {Promise<Object|null>} Face detection or null
 */
export async function detectLargestFace(imageBuffer) {
  const faces = await detectFaces(imageBuffer);

  if (faces.length === 0) return null;

  // Find largest face by area
  return faces.reduce((largest, face) => {
    const area = face.box.width * face.box.height;
    const largestArea = largest.box.width * largest.box.height;
    return area > largestArea ? face : largest;
  });
}

/**
 * Extract face region from image with padding
 * @param {Buffer} imageBuffer - Image buffer
 * @param {Object} box - Face bounding box
 * @param {number} padding - Padding ratio (0.3 = 30%)
 * @returns {Promise<{buffer: Buffer, box: Object}>}
 */
export async function extractFaceRegion(imageBuffer, box, padding = 0.3) {
  const img = await loadImage(imageBuffer);

  const padW = Math.round(box.width * padding);
  const padH = Math.round(box.height * padding);

  const x = Math.max(0, box.x - padW);
  const y = Math.max(0, box.y - padH);
  const width = Math.min(img.width - x, box.width + 2 * padW);
  const height = Math.min(img.height - y, box.height + 2 * padH);

  const canvas = createCanvas(width, height);
  const ctx = canvas.getContext('2d');
  ctx.drawImage(img, x, y, width, height, 0, 0, width, height);

  return {
    buffer: canvas.toBuffer('image/png'),
    box: { x, y, width, height },
  };
}

/**
 * Create an elliptical mask for face blending
 * @param {number} width - Mask width
 * @param {number} height - Mask height
 * @returns {Buffer} Grayscale mask buffer
 */
export function createFaceMask(width, height) {
  const canvas = createCanvas(width, height);
  const ctx = canvas.getContext('2d');

  // Create radial gradient for smooth edges
  const centerX = width / 2;
  const centerY = height / 2;
  const radiusX = width * 0.45;
  const radiusY = height * 0.48;

  // Draw ellipse with gradient
  ctx.fillStyle = 'black';
  ctx.fillRect(0, 0, width, height);

  ctx.beginPath();
  ctx.ellipse(centerX, centerY, radiusX, radiusY, 0, 0, Math.PI * 2);
  ctx.fillStyle = 'white';
  ctx.fill();

  return canvas.toBuffer('image/png');
}

export default {
  loadModels,
  detectFaces,
  detectLargestFace,
  extractFaceRegion,
  createFaceMask,
};
